README
                       TcosMonitor
               (mariodebian AT gmail DOT com)

  TcosMonitor is a pygtk app to control and supervise
  a thin client classroom based on TCOS, PXES, LTSP or others.


  The project have 2 different parts:

   * Teacher Tool (TcosMonitor) made with pygtk and other
     standar libs

   * Thin Client daemon (tcosxmlrpc) developed in C based on
     xmlrpc libs and abyss web server code and libs. 
     For debug output exec TCOS_DEBUG=1 /usr/bin/tcosxmlrpc /etc/abys.conf

   * Some extra tools (hex2ascii, update-tcospasswd, etc...)


  The initial code is written from scratch helped with pygtk
  tutorial [1] and some web pages (blogs, maillist, etc...)

  I have reutilized some python/C code from others:
  
    ping.py
       High based on yetist [2], I have added a second
       class named PingPort writen by me. 
         (don't know license of it, japanesse page)
  
    xmlrpc/hex2ascii.c 
       Based on http://www.bigbold.com/snippets/posts/show/2073
       this is a small tool to convert an HEX string into ASCII
       used by TCOS thin client when DHCP server is configured
       with no static MAC.
        (web says:
          "Snippets is a public source code repository."
            I suppose that I can use without problems )


    TcosDbus.py class
        Based on dbus bindings of Student-Control-Center of Ubuntu [3]
           GPL Copyright Holder: Oliver Grawert <ogra@ubuntu.com>




Links:
[1] http://www.pygtk.org/pygtk2tutorial-es/
[2] http://www.osprg.org/modules/newbb/viewtopic.php?forum=3&post_id=487
[3] http://people.ubuntu.com/~ogra/bzr-archive/student-control-panel-0.1/



Changelog
tcosmonitor (0.1.18.1) unstable; urgency=low

  * Move dbus conf file and tcos-dbus-client to tcos-tcosmonitor

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 07 Jul 2007 14:02:41 +0200

tcosmonitor (0.1.18.0) unstable; urgency=low

  * Remove pyexamples dir
  * xmlrpc/sh/restartx.sh:
    - call restartxorg instead of killing X
  * tcos-devices-ng:
    - Allot to disable floppy cdroms or hdd from conf file
      (disable_floppy=1, disable_cdroms=1, disable_hdd=1)
  * xmlrpc/*:
    - Add tcos.who method to retrieve username connected at :0

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 07 Jul 2007 13:28:15 +0200

tcosmonitor (0.1.17.2) unstable; urgency=low

  * xmlrpc/Makefile xmlrpc/var/etc/abyss.conf:
    - Set correct paths for tcos-standalone and tcos thin client

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 27 Jun 2007 23:54:22 +0200

tcosmonitor (0.1.17.1) unstable; urgency=low

  * New package tcos-standalone (Use tcosmonitor with not standalone hosts)
  * xmlrpc/*.h:
    - Add TCOS_PATH define to tcosxmlrpc, to allow running external tools from other paths
  * xmlrpc/sh/*:
    - Ad conditions to exec or not some tools.
  * TcosTrayIcon.py:
    - Remove StatusIcon declaration (break at gtk < 2.10)
  * Add support of standalone mode in TcoaActions.py
  * TcosActions.py:
    - patch to start/stop video broadcast (vidal_joshur at gva dot es)
    - patch to send files with rsync (vidal_joshur at gva dot es)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 27 Jun 2007 20:10:18 +0200

tcosmonitor (0.1.16.2) unstable; urgency=low

  * udev/tcos-udevd.sh:
    - Force udev to save ID_FS_TYPE in disk, not only partitions

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 20 Jun 2007 21:18:12 +0200

tcosmonitor (0.1.16.1) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Add support to mount NTFS with ntfs-3g driver
  * tcos-devices-ng.py:
    - Mount ntfs-3g in a thread (fuse mounts exec a process)
  * TcosXmlRpc.py:
    - Don't fail in GetDevicesInfo (use try--except)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Wed, 20 Jun 2007 15:14:16 +0200

tcosmonitor (0.1.16) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Fix typo in include header
  * images/*cdrom*
    - new image based on gnome theme
  * tcos-devices-ng.py:
    - Fix translation strings (vars go outside _('%s')%var )
    - Add umount_all function called from exit
    - Fix USB device remount
    - Allow to specify mounts points with a new conf file ~/.tcos-devices-ng.conf
  * Add new man page: man/tcos-devices-ng.1

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 19 Jun 2007 18:34:18 +0200

tcosmonitor (0.1.15.5) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Option gethdd now returns all type disk except swap or iso9660

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Tue, 19 Jun 2007 12:40:51 +0200

tcosmonitor (0.1.15.4) unstable; urgency=low

  * xmlrpc/sh/devicesctl.sh:
    - Add --gethdd param and return ext3 or vfat devices in /etc/fstab
  * tcos-devices-ng.py:
    - Read hdd device partitions in thin client and put in menu
  * images/hdd_*mount.png:
    - Add hdd icons
  * po/*:
    - Update translations (new hdd strings)

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 18 Jun 2007 18:52:00 +0200

tcosmonitor (0.1.15.3) unstable; urgency=low

  * lockscreen/* :
    - Change copyright header
    - Compile with imlib-config flags and libs
  * dbus/81tcos-utils:
    - Start tcos-devices --daemon and tcos-devices --hidden
  * tcos-devices-ng.py:
    - Don't use shared vars before importing module shared

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Mon, 18 Jun 2007 11:06:10 +0200

tcosmonitor (0.1.15.2) unstable; urgency=low

  * conf/tcos.conf man/* :
    - fix some english typos (close ticket #3) thanks to Andres Jimenez again
  * lockscreen/*:
    - Add support to put a png/jpg, xpm, etc... image in lockscreen
    - Default image locked.png
  * tcos-devices-ng.py:
    - Clean code and remote extra comments
  * pyexamples:
    - Add some python examples

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sun, 17 Jun 2007 17:01:00 +0200

tcosmonitor (0.1.15.1) unstable; urgency=low

  * shared.py:
    - Remove some ugly debug info
  * xmlrpc/sh/devicesctl.sh:
    - Add --exists parameter to know if a file (or device) exists
  * xmlrpc/sh/soundctl.sh:
    - Exis if not soundcard found
  * xmlrpc/xauth.c:
    - Compare local hostname with cookie hostname.
  * tcos-devices.py:
    - Exit if X cookie don't match
  * images/*:
    - Add some tcos-devices-ng.py images
  * TcosTrayIcon.py:
    - New Tcos class (used to build systray menu)
  * tcos-devices-ng.py:
    - New and more powerfull remote device manager
  * Rename dbus/81tcos-dbus-client to dbus/81tcos-utils and start tcos-devices from there
  * po/*:
    - Update spanish translation

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Thu, 14 Jun 2007 21:15:15 +0200

tcosmonitor (0.1.14.4) unstable; urgency=low

  * TcosXauth.py:
    - Read DISPLAY info and search for hostname and IP address

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat,  9 Jun 2007 19:33:04 +0200

tcosmonitor (0.1.14.3) unstable; urgency=low

  * tcos-devices.py, tcos-volume-manager.py:
    - If gtk version is greater than 2.10 use gtk.statusicon instead of egg.trayicon

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 09 Jun 2007 17:11:33 +0200

tcosmonitor (0.1.14.2) unstable; urgency=low

  * TcosXauth.py:
    - Read xauth cookie with -n param to not resolve hostname
    - Try (with try: except: to get hostname with socket) not xauth locks

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Fri, 08 Jun 2007 22:15:23 +0200

tcosmonitor (0.1.14.1) unstable; urgency=low

  * Initialize.py:
    - Fix AboutDialog, close and destry button for gtk > 2.10

shared.py
# -*- coding: UTF-8 -*-
##########################################################################
# TcosMonitor writen by MarioDebian <mariodebian@gmail.com>
#
#    TcosMonitor version __VERSION__
#
# Copyright (c) 2006 Mario Izquierdo <mariodebian@gmail.com>
# All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
###########################################################################

# Defaults values of TcosMonitor


import os
from gettext import gettext as _
from gettext import bindtextdomain, textdomain
from locale import setlocale, LC_ALL

have_display=False
allow_local_display=False

# import if DISPLAY is defined
if "DISPLAY" in os.environ:
    if os.environ["DISPLAY"] != "":
        have_display=True
        import pygtk
        pygtk.require('2.0')
        import gtk

# program name to use in gettext .mo
PACKAGE = "tcosmonitor"



# version
version="__VERSION__"
website="http://www.tcosproject.org"
website_label=_("TcosMonitor web page")


# default debug value (overwrite with --debug or -d)
debug=False

# default TCOS config file (default in this path, if installed use global)
tcos_config_file="./tcos.conf"

# if exec from svn or sources dir
if os.path.isdir('./debian') and  os.path.isdir('./xmlrpc'):
    LOCALE_DIR = "./po/"
    GLADE_DIR = "./"
    IMG_DIR = "./images/"
    tcos_config_file="./tcos.conf"
else:
    tcos_config_file="/etc/tcos/tcos.conf"
    GLADE_DIR = "/usr/share/tcosmonitor/"
    IMG_DIR = "/usr/share/tcosmonitor/images/"
    LOCALE_DIR = "/usr/share/locale"


# config file
config_file=os.path.expanduser('~/.tcosmonitor.conf')

scan_methods=[
"netstat", 
"ping"
]


DefaultConfig=[
["populate_list_at_startup", 0, "int"],
["work_as_cyber_mode", 0, "int"],
["refresh_interval", 10, "int"],
["cache_timeout", 0, "int"],
["scan_network_method", "netstat", "str"],
["scrot_size", 65, "int"],      # % of screenshot
["miniscrot_size", 25, "int"],      # % of screenshot
["xmlrpc_username", "user", "str"], 
["xmlrpc_password","", "str"], 
["network_interface","eth0", "str"], 
["tcosinfo", 1, "int"], 
["cpuinfo", 0, "int"],
["kernelmodulesinfo", 0, "int"], 
["pcibusinfo", 0, "int"], 
["ramswapinfo", 0, "int"],
["processinfo", 0, "int"], 
["networkinfo", 0, "int"], 
["xorginfo", 0, "int"], 
["soundserverinfo", 0, "int"], 
["systemprocess", 0, "int"], 
["tcosmonitorversion", version, "str"]
]
# method ping is list 0 of combo_scan_method


# gettext support
setlocale( LC_ALL )
bindtextdomain( PACKAGE, LOCALE_DIR )
textdomain( PACKAGE )

# text file enabling or disabling tcos-devices or tcos-volume-manager
module_conf_file="/etc/tcos/tcosmonitor.conf"

httpd_port=8081

xmlremote_port=8080
xmlremote_url="/RPC2"

pulseaudio_soundserver_port=4713
sound_only_channels=["Master", "PCM", "Line", "CD", "Mic", "Aux", "vol", "pcm", "line", "cd", "mic"]


cache_timeout=20

wait_between_many_host=0.1
socket_default_timeout=15

dbus_disabled=False

disable_textview_on_update=False


NO_LOGIN_MSG="----"

##
##  entry completion example apps
##

appslist=[
'xterm', 
'firefox', 
'iceweasel',
'gimp', 
'oowriter', 
'oocalc', 
'oodraw' ,
'gnumeric',
'abiword' ,
'gedit', 
'jclic', 
'amsn', 
'gftp',
'gcalctool',
'nautilus',
'konqueror',
'beep-media-player',
'audacious',
'tcos-volume-manager',
'tcos-devices',
'tcos-devices-ng'
]

###
###   [ TEXT, ICON (in images dir) ]
###
onehost_menuitems=[
 [ _("Refresh terminal info"), "menu_refresh.png"] ,          #action=0
 [ _("Clean info about terminal"), "menu_clear.png"] ,        #action=1
 [ _("Reboot"), "menu_reboot.png"] ,                          #action=2
 [ _("Poweroff"), "menu_poweroff.png"] ,                      #action=3
 [ _("Lock screen"), "menu_lock.png" ] ,                      #action=4
 [ _("Unlock screen"), "menu_unlock.png" ] ,                  #action=5
 [ _("Connect to remote screen (iTALC)"), "menu_remote.png" ],#action=6
 [ _("Connect to remote screen (VNC)"), "menu_remote.png" ] , #action=7
 [ _("Screenshot"), "menu_screenshot.png" ] ,                 #action=8
 [ _("Give a remote xterm"), "menu_xterm.png" ]  ,            #action=9
 [ _("Configure this host"), "menu_configure.png" ] ,         #action=10
 [ _("Logout client"),  "menu_restartx.png" ] ,               #action=11
 [ _("Restart X session with new settings"), "menu_newconf.png" ] , #action=12
 [ _("Exec app on user display") , "menu_exec.png" ] ,        #action=13
 [ _("Send a text message to user") , "menu_msg.png" ] ,     #action=14
 [ _("Show running apps of this client") , "menu_proc.png" ], #action=15
 ]


allhost_menuitems=[
 [ _("Reboot all clients"), "menu_reboot.png"] ,                  #action=0
 [ _("Poweroff all clients"), "menu_poweroff.png"] ,              #action=1
 [ _("Lock all screens"), "menu_lock.png" ] ,                     #action=2
 [ _("Unlock all screens"), "menu_unlock.png" ] ,                 #action=3
 [ _("Logout clients"), None ] ,   # FIXME need an icon           #action=4
 [ _("Restart X session of all clients"),  "menu_restartx.png" ] ,#action=5
 [ _("Exec same app in all connected users") , "menu_exec.png" ] ,#action=6
 [ _("Send a text message to all connected users") , "menu_msg.png" ], #action=7
 [ _("Enter demo mode, all connected users see my screen") , "menu_remote.png" ], #action=8
 [ _("Exit demo mode") , "menu_remote.png" ],                     #action=9
 [ _("Capture All clients screens") , "menu_screenshot.png" ],    #action=10
 [ _("Video broadcast") , "menu_remote.png" ],                    #action=11
 [ _("Stop Video broadcast") , "menu_remote.png" ],               #action=12
 [ _("Send files") , "menu_remote.png" ]                          #action=13
 ]


# this list contains all process to not show in user processes
system_process=[
 "dbus-daemon",
 "tcos-dbus-client", 
 "scp-client", 
 "ssh-agent", 
 "dbus-launch", 
 "gam_server", 
 "gconfd", 
 "gnome-keyring-daemon", 
 "gnome-applets", 
 "gnome-pty-helper", 
 "gnome-settings-daemon", 
 "-applet",
 "gnome-vfs",  
 "panel-plugins",
 "dcop", 
 "bonobo",
 "xauth",
 "tcos-devices",
 "tcos-volume-manager",
 "notification-daemon",
 "bash"
]



# TcosPersonalize stuff

remotehost=None

xsession_values=[
"XDMCP",  
"local", 
"sshX",
"FreeNX"
"rDesktop"
]
xsession_default="XDMCP"

# FIXME need to support other drivers
xdriver_values=[
"auto"
]
xdriver_default="auto"


# obtained from configure xorg
# grep "@" bin/configurexorg | \
#    grep Hz | awk -F ")" '{print $1}' | awk '{print $1", "}'
xres_values=[
"640x480 @ 60Hz",
"640x480 @ 72Hz",
"800x600 @ 60Hz",
"800x600 @ 72Hz",
"800x600 @ 85Hz",
"832x624 @ 75Hz",
"1024x768 @ 60Hz", # <= default
"1024x768 @ 70Hz",
"1024x768 @ 75Hz",
"1152x768 @ 54.8Hz",
"1152x864 @ 60Hz",
"1152x864 @ 75Hz",
"1280x768 @ 60Hz",
"1280x800 @ 60Hz",
"1280x960 @ 60Hz",
"1280x960 @ 85Hz",
"1280x1024 @ 60Hz",
"1400x1050 @ 60Hz",
"1400x1050 @ 75Hz",
"1600x1024 @ 85Hz",
"1600x1200 @ 60Hz",
"1600x1200 @ 75Hz",
"1600x1200 @ 85Hz",
"1680x1050 @ 60Hz",
"1792x1344 @ 75Hz",
"1792x1344 @ 60Hz",
"1856x1392 @ 60Hz",
"1856x1392 @ 75Hz",
"1920x1200 @ 60Hz",
"1920x1440 @ 60Hz",
"1920x1440 @ 75Hz",
"1920x1440 @ 85Hz",
"2048x1536 @ 60Hz",
"2048x1536 @ 75Hz",
"2048x1536 @ 85Hz",
]
xres_default="1024x768 @ 60Hz"


xdepth_values=[
"24", 
"16", 
"15", 
"8",
"4",  
"1"
]
xdepth_default="16"



PersonalizeConfig=[
["xdriver", xdriver_default, "str"],
["xres", xres_default, "str"],
["xdepth", xdepth_default, "str"], 
["xmousewheel",1, "int"], 
["xdontzap",0, "int"], 
["xdpms",1, "int"], 
["xsession", xsession_default, "str"], 
["xhorizsync", "", "str"], 
["xvertsync", "", "str"], 
["tcospersonalizeversion", version, "str"]
]



# shared functions

def print_debug(txt):
    if debug:
        print "%s::%s" %(__name__, txt)
    return


if have_display:
    def ask_msg(txt):
        response="yes"
        d = gtk.MessageDialog(None,
            gtk.DIALOG_MODAL |
            gtk.DIALOG_DESTROY_WITH_PARENT,
            gtk.MESSAGE_QUESTION,
            gtk.BUTTONS_YES_NO,
            txt)
        if d.run() == gtk.RESPONSE_YES:
            response=True
        else:
            response=False
        d.destroy()
        mydict={}
        print_debug( _("QUESTION: %(txt)s, RESPONSE %(response)s")  %{"txt":txt, "response":response} )
        return response

    def info_msg(txt):
        d = gtk.MessageDialog(None,
                      gtk.DIALOG_MODAL |
                      gtk.DIALOG_DESTROY_WITH_PARENT,
                      gtk.MESSAGE_INFO,
                      gtk.BUTTONS_OK,
                      txt)
        d.run()
        d.destroy()
        print_debug ( _("INFO: %s") % txt )

    def error_msg(txt):
        d = gtk.MessageDialog(None,
                      gtk.DIALOG_MODAL |
                      gtk.DIALOG_DESTROY_WITH_PARENT,
                      gtk.MESSAGE_WARNING,
                      gtk.BUTTONS_OK,
                      txt)
        d.run()
        d.destroy()
        print_debug ( _("ERROR: %s") % txt )


    def test_start(module):
        """
            read conf file and test if module is active
        """
        f=open(module_conf_file, "r")
        conf=f.readlines()
        f.close()
        
        for line in conf:
            if line == '\n': continue
            if line.find('#') == 0: continue
            line=line.replace('\n', '')
            if "=" in line:
                if line.split('=')[0] == module:
                    if line.split('=')[1] == "0":
                        return False
                    else:
                        return True


from threading import Thread


class Workers:
    def __init__(self, main, target, args, dog=True):
        self.dog=dog
        self.main=main
        self.target=target
        self.args=args
        
        if not self.dog:
            #print_debug ( "worker() no other jobs job=%s args=%s" %(self.target, self.args) )
            self.th=Thread(target=self.target, args=(self.args) )
            self.__stop=True
            return
        
        if self.main.worker_running == True:
            #print_debug ( "worker() other jobs pending NO START job=%s args=%s" %(self.target, self.args) )
            pass
        else:
            #print_debug ( "worker() no other jobs job=%s args=%s" %(self.target, self.args) )
            self.th=Thread(target=self.target, args=(self.args) )
            self.__stop=True
    
    def start_watch_dog(self, dog_thread):
        if not self.dog:
            #print_debug ( "start_watch_dog() dog DISABLED" )
            return
        print_debug ( "start_watch_dog() starting watch dog..." )
        watch_dog=Thread(target=self.watch_dog, args=([dog_thread]) )
        watch_dog.start()

    def watch_dog(self, dog_thread):
        print_debug ( "watch_dog()  __init__ " )
        dog_thread.join()
        self.set_finished()
        print_debug ( "watch_dog() FINISHED" )
        
    def start(self):
        try:
            self.main.progressbutton.set_sensitive(True)
        except:
            pass
        
        if self.main.worker_running == False:
            self.th.start()     # start thread
            self.set_started()  # config var as started
            self.start_watch_dog(self.th) # start watch_dog
        else:
            print_debug ( "worker() other work pending... not starting" )
        
    def stop(self):
        self.__stop=True
        self.__finished=True
        #self.main.worker_running=False
        
    def set_finished(self):
        self.__finished = True
        self.__stop=False
        self.main.worker_running=False

    def set_started(self):
        self.__finished=False
        self.__stop=False
        self.main.worker_running=True

    def is_stoped(self):
        return self.__stop
        
    def get_finished(self):
        return self.__finished

    def set_for_all_action(self, function, allhost, action):
        action_args=[allhost, action]
        Thread( target=function, args=(action_args) ).start()


